rskj-core/src/main/java/co/rsk/trie/MultiTrieStore.java

```diff
@@ -0,0 +1,196 @@
+     */
+    @Override
+    public synchronized void save(Trie trie) {
+        save(trie, true);
+    }
+
+    /**
+     * @param forceSaveRoot allows saving the root node even if it's embeddable
+     */
+    private void save(Trie trie, boolean forceSaveRoot) {
+        // Invariant: a trie saved in epoch N is fully contained in d(N) and d(N - 1).
+        // This ensures all epoch N tries survive after discarding d(N - 2).
+        if (epochNSavedTries.contains(trie) || epochN1SavedTries.contains(trie)) {
+            // TODO(mc): can we return? could its children be in d(N - 2)?
+            return;
+        }
+
+        trie.getLeft().getNode().ifPresent(t -> save(t, false));
+        trie.getRight().getNode().ifPresent(t -> save(t, false));
+
+        if (trie.hasLongValue()) {
+            epochN.put(trie.getValueHash().getBytes(), trie.getValue());
+        }
+
+        if (trie.isEmbeddable() && !forceSaveRoot) {
+            return;
+        }
+
+        epochN.put(trie.getHash().getBytes(), trie.toMessage());
+        epochNSavedTries.add(trie);
+    }
+
+    @Override
+    public synchronized Trie retrieve(byte[] hash) {
+        byte[] epochNData = epochN.get(hash);
+        if (epochNData != null) {
+            Trie trie = Trie.fromMessage(epochNData, this);
+            epochNSavedTries.add(trie);
+            return trie;
+        }
+
+        byte[] epochN1Data = epochN1.get(hash);
+        if (epochN1Data != null) {
+            Trie trie = Trie.fromMessage(epochN1Data, this);
+            epochN1SavedTries.add(trie);
+            return trie;
+        }
+
+        byte[] epochN2Data = epochN2.get(hash);
+        if (epochN2Data != null) {
+            Trie trie = Trie.fromMessage(epochN2Data, this);
+            epochN2SavedTries.add(trie);
+            return trie;
+        }
+
+        throw new IllegalArgumentException(String.format(
+                "The trie with root %s is missing in this store", Hex.toHexString(hash)
+        ));
+    }
+
+    @Override
+    public synchronized byte[] retrieveValue(byte[] hash) {
+        byte[] epochNData = epochN.get(hash);
+        if (epochNData != null) {
+            return epochNData;
+        }
+
+        byte[] epochN1Data = epochN1.get(hash);
+        if (epochN1Data != null) {
+            return epochN1Data;
+        }
+
+        return epochN2.get(hash);
+    }
+
+    @Override
+    public synchronized void flush() {
+        epochN.flush();
+    }
+
+    // TODO(mc) collect in the background
+    @Override
+    public synchronized void collect(Trie lastFrontierTrie, long lastEpoch) {
+        // TODO(mc) could it be greater?
+        if (lastEpoch < currentEpoch) {
+            logger.warn("Skipping collection with epoch {} lower than current epoch {}", lastEpoch, currentEpoch);
+            return;
+        }
+
+        // TODO(mc) is this necessary with the save invariant?
+        ensureTrieForCollection(lastFrontierTrie, true);
```

Comment:
I think it is needed. d(N) and d(N-1) contain the full current state, a merge is required to generate the new d(N-1) storage.